src/librustc_infer/infer/combine.rs

```diff
@@ -719,3 +721,72 @@ fn float_unification_error<'tcx>(
     let (ty::FloatVarValue(a), ty::FloatVarValue(b)) = v;
     TypeError::FloatMismatch(ty::relate::expected_found_bool(a_is_expected, &a, &b))
 }
+
+struct ConstOccursChecker<'cx, 'tcx> {
+    infcx: &'cx InferCtxt<'cx, 'tcx>,
+
+    /// The vid of the const variable that is in the process of being
+    /// instantiated; if we find this within the const we are folding,
+    /// that means we would have created a cyclic const.
+    target_vid: ty::ConstVid<'tcx>,
+
+    cycle: bool,
+}
+
+impl<'cx, 'tcx> ConstOccursChecker<'cx, 'tcx> {
+    fn check(mut self, t: &'tcx ty::Const<'tcx>) -> RelateResult<'tcx, ()> {
+        self.visit_const(t);
+        if self.cycle { Err(TypeError::CyclicConst(t)) } else { Ok(()) }
+    }
+}
+
+impl TypeVisitor<'tcx> for ConstOccursChecker<'_, 'tcx> {
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+        debug!("ConstOccursChecker: t={:?}", t);
+
+        match t.kind {
+            ty::Infer(ty::TyVar(vid)) => {
+                let vid = self.infcx.inner.borrow_mut().type_variables().root_var(vid);
+                let probe = self.infcx.inner.borrow_mut().type_variables().probe(vid);
+                match probe {
+                    TypeVariableValue::Known { value: u } => {
+                        debug!("ConstOccursChecker: known value {:?}", u);
+                        self.visit_ty(u)
+                    }
+                    TypeVariableValue::Unknown { universe: _ } => {
+                        // FIXME: We may have to do something similar to what
+                        // `Generalizer` is doing here.
+                        false
```

Comment:
I am not too familiar with what has to be done with `UniverseIndex` here, when instantiating type inference variables,
we sometimes adjust them, which we might also have to do for consts

https://github.com/rust-lang/rust/blob/0cd7ff7ddfb75a38dca81ad3e76b1e984129e939/src/librustc_infer/infer/combine.rs#L660-L668

and

https://github.com/rust-lang/rust/blob/0cd7ff7ddfb75a38dca81ad3e76b1e984129e939/src/librustc_infer/infer/combine.rs#L558-L587

Reply:
OK, so, the idea is that... if we are providing a value `V` for some variable `?A`, then we have to ensure that the value `V` can be named from within the universe of `?A`. e.g., if `?A` is in the root universe, then `V` can only reference names in the root universe.

Therefore, if `V` contains some other variable `?B`, then the universe `Universe(?B)` of `?B` must be some ancestor of `Universe(?A)`.

We have to decide how "eagerly" we want to report these errors.

In chalk at least, we treat type variables and lifetime variables differently. For example, if we have `exists<T> { forall<U> { T = U } }` in chalk, we consider that to be a unification failure.  But if we have `exists<'a> { forall<'b> { 'a = 'b } }` in chalk, we produce a region constraint that will never be satisfied.

However, Rust doesn't presently have binders over type variables like this, so the question doesn't arise in stable rust. I guess I have to go read what the unification code does now.

Reply:
Hm, so I guess the goal is to use a `TypeFolder` instead of a visitor here and update the type and const variables if they are in a universe which can't be named.

I am still not completely sure where this is needed and how to test this, so it might be fine to merge this with a FIXME here

Reply:
Given the limited range of "forall" currently possible in Rust, that may well be true...not sure. I guess there might be things like `exists<'a> { forall<'b> { <T as Foo<'a>>::BAR = <T as Foo<'b>>::BAR } }` or something that might indirectly relate lifetimes with universes and be led astray.

Reply:
I guess I should review the type unification code.